service/serviceUtils.go

```diff
@@ -0,0 +1,211 @@
+		return &harvestType, nil
+	case "public_services":
+		harvestType := model.PublicServiceHarvestType
+		return &harvestType, nil
+	case "events":
+		harvestType := model.EventHarvestType
+		return &harvestType, nil
+	}
+
+	return nil, errors.New(string(routingKeyPrefix) + " is not a valid harvest type")
+}
+
+func ReasonedOrIngestedReport(routingKey string, startAndEndTime model.StartAndEndTime) (*model.HarvestReport, error) {
+	splitKey := strings.Split(routingKey, ".")
+	if len(splitKey) != 2 {
+		return nil, errors.New(string(routingKey) + " is not a valid routing key")
+	}
+
+	harvestType, err := HarvestTypeFromRoutingKey(splitKey[0])
+	if err != nil {
+		return nil, err
+	}
+
+	report := model.HarvestReport{
+		Id:               splitKey[1],
+		Url:              nil,
+		DataType:         model.HarvestTypeEnum(*harvestType),
+		HarvestError:     false,
+		StartTime:        startAndEndTime.StartTime,
+		EndTime:          startAndEndTime.EndTime,
+		ErrorMessage:     nil,
+		ChangedCatalogs:  []model.FdkIdAndUri{},
+		ChangedResources: []model.FdkIdAndUri{},
+	}
+
+	return &report, nil
+}
+
+func calculateHarvestStatusesFromReports(
+	harvestReports model.HarvestReports,
+	reasoningReports model.HarvestReports,
+	ingestReports model.HarvestReports,
+) (*model.HarvestStatuses, error) {
+	var statusList []model.HarvestStatus
+	for _, report := range harvestReports.Reports {
+		reasoningReport, _ := reasoningReports.Reports[string(report.DataType)]
+		ingestReport, _ := ingestReports.Reports[string(report.DataType)]
+		status, err := harvestStatusFromRelevantReports(report, &reasoningReport, &ingestReport)
+		if err != nil {
+			return nil, err
+		}
+		statusList = append(statusList, *status)
+	}
+
+	harvestStatuses := model.HarvestStatuses{
+		Id:       harvestReports.Id,
+		Statuses: statusList,
+	}
+
+	return &harvestStatuses, nil
+}
+
+func harvestStatusFromRelevantReports(
+	harvestReport model.HarvestReport,
+	reasoningReport *model.HarvestReport,
+	ingestReport *model.HarvestReport,
+) (*model.HarvestStatus, error) {
+	if harvestReport.HarvestError {
+		status := model.HarvestStatus{
+			HarvestType:  string(harvestReport.DataType),
+			Status:       model.HarvestError,
+			ErrorMessage: harvestReport.ErrorMessage,
+			StartTime:    harvestReport.StartTime,
+			EndTime:      &harvestReport.EndTime,
+		}
+
+		return &status, nil
+	}
+
+	harvestIsInProgress, err := IsInProgress(harvestReport, reasoningReport, ingestReport)
+	if err != nil {
+		return nil, err
+	}
+
+	if harvestIsInProgress {
+		status := model.HarvestStatus{
+			HarvestType: string(harvestReport.DataType),
+			Status:      model.HarvestInProgress,
+			StartTime:   harvestReport.StartTime,
+		}
+		return &status, nil
+	} else {
+		status := model.HarvestStatus{
+			HarvestType: string(harvestReport.DataType),
+			Status:      model.HarvestDone,
+			StartTime:   harvestReport.StartTime,
+			EndTime:     &ingestReport.EndTime,
+		}
+		return &status, nil
+	}
```

Comment:
Om feltene som er like i `if` og `else` samles på toppen, ser man tydeligere hva som er likt og ulikt i disse to tilfellene. Det er her også litt kortere, mindre duplisert og bare én return :smile: 
```suggestion
	status := model.HarvestStatus{
		HarvestType: string(harvestReport.DataType),
		StartTime:   harvestReport.StartTime,
	}
	if harvestIsInProgress {
		status.Status = model.HarvestInProgress
	} else {
		status.Status = model.HarvestDone
	        status.EndTime = &ingestReport.EndTime
	}
	return &status, nil
```

Reply:
668a963